### PR TITLE
remove dependency of vimproc

### PR DESCRIPTION
**Pull request by mattn**
tsuquyomi

### CODE REVIEW

**Comment by Quramy** [comment]
Hi @mattn,  thanks for your PR!  

To attempt this feature, I tried to exec `:echo tsuquyomi#hint()` and I got the following message:

```
"test/tsClient/vest/resources/samplePrjs/prj001/main.ts" 5L, 80C
Error detected while processing function tsuquyomi#hint[2]..tsuquyomi#tsClient#tsQuickinfo[2]..tsuquyomi#tsClient#sendCommandSyncResponse[3]..tsuquyomi#tsClient#sendRequest:
line    3:
E631: ch_sendraw(): write failed
```

And `:TsuStatusServer` indicated "dead". Do you have some work-arounds?

My env:

```
VIM - Vi IMproved 8.0 (2016 Sep 12, compiled Jan  2 2017 00:45:39)
MacOS X (unix) version
Included patches: 1-104
Compiled by kurami@kurami-2.local
Huge version without GUI.  Features included (+) or not (-):
+acl             +file_in_path    +mouse_sgr       +tag_old_static
+arabic          +find_in_path    -mouse_sysmouse  -tag_any_white
+autocmd         +float           +mouse_urxvt     -tcl
-balloon_eval    +folding         +mouse_xterm     +termguicolors
-browse          -footer          +multi_byte      +terminfo
++builtin_terms  +fork()          +multi_lang      +termresponse
+byte_offset     -gettext         -mzscheme        +textobjects
+channel         -hangul_input    +netbeans_intg   +timers
+cindent         +iconv           +num64           +title
-clientserver    +insert_expand   +packages        -toolbar
+clipboard       +job             +path_extra      +user_commands
+cmdline_compl   +jumplist        -perl            +vertsplit
+cmdline_hist    +keymap          +persistent_undo +virtualedit
+cmdline_info    +lambda          +postscript      +visual
+comments        +langmap         +printer         +visualextra
+conceal         +libcall         +profile         +viminfo
+cryptv          +linebreak       -python          +vreplace
+cscope          +lispindent      -python3         +wildignore
+cursorbind      +listcmds        +quickfix        +wildmenu
+cursorshape     +localmap        +reltime         +windows
+dialog_con      -lua             +rightleft       +writebackup
+diff            +menu            -ruby            -X11
+digraphs        +mksession       +scrollbind      -xfontset
-dnd             +modify_fname    +signs           -xim
-ebcdic          +mouse           +smartindent     -xpm
+emacs_tags      -mouseshape      +startuptime     -xsmp
+eval            +mouse_dec       +statusline      -xterm_clipboard
+ex_extra        -mouse_gpm       -sun_workshop    -xterm_save
+extra_search    -mouse_jsbterm   +syntax          
+farsi           +mouse_netterm   +tag_binary      
   system vimrc file: "$VIM/vimrc"
     user vimrc file: "$HOME/.vimrc"
 2nd user vimrc file: "~/.vim/vimrc"
      user exrc file: "$HOME/.exrc"
       defaults file: "$VIMRUNTIME/defaults.vim"
  fall-back for $VIM: "/Users/yosuke/local/bin/share/vim"
Compilation: gcc -c -I. -Iproto -DHAVE_CONFIG_H   -DMACOS_X_UNIX  -g -O2 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1       
Linking: gcc   -L/usr/local/lib -o vim        -lm -lncurses  -liconv -framework Cocoa           
```

File: autoload/tsuquyomi/config.vim [hunk 1]
```diff
@@ -105,12 +109,21 @@ function! tsuquyomi#config#tsscmd()
   return l:cmd
 endfunction
 
+function! s:system(cmd)
+  let out = ''
+  let job = job_start([&shell, &shellcmdflag, a:cmd], {'out_cb': {ch,msg->[execute("let out .= msg"), out]}, 'out_mode': 'raw'})
```
Comment (prabirshrestha) [inline autoload/tsuquyomi/config.vim]: seems to me like it should now be easy to also support neovim?

File: autoload/tsuquyomi/config.vim [hunk 2]
```diff
@@ -21,11 +19,11 @@ let s:tss_version = {'is_valid': 0, 'out': '???'}
 function! tsuquyomi#config#preconfig()
 
   if !exists('g:tsuquyomi_is_available')
-    if !s:P.is_available()
-      " 1. vimproc installation check
+    if !has('patch-8.0.0')
```
Comment (prabirshrestha) [inline autoload/tsuquyomi/config.vim]: might be also check for job and lambda?

File: autoload/tsuquyomi/config.vim [hunk 3]
```diff
@@ -21,11 +19,11 @@ let s:tss_version = {'is_valid': 0, 'out': '???'}
...
+      " 1. vim8 installation check
       let g:tsuquyomi_is_available = 0
       call s:deleteCommand()
-      echom '[Tsuquyomi] Shougo/vimproc.vim is not installed. Please install it.'
+      echom '[Tsuquyomi] vim8 is not installed. Please install it.'
```
Comment (prabirshrestha) [inline autoload/tsuquyomi/config.vim]: vim8 with jobs and lambda is not installed?

File: ftplugin/typescript.vim [hunk 4]
```diff
@@ -11,7 +11,7 @@ endif
 
 if(!exists('g:tsuquyomi_is_available'))
   let g:tsuquyomi_is_available = 0
-  echom '[Tsuquyomi] Shougo/vimproc.vim is not installed. Please install it.'
+  echom '[Tsuquyomi] vim8 is not installed. Please install it.'
```
Comment (prabirshrestha) [inline ftplugin/typescript.vim]: vim8 with jobs and lambda is not installed?

File: ftplugin/javascript.vim [hunk 5]
```diff
@@ -15,7 +15,7 @@ endif
 
 if(!exists('g:tsuquyomi_is_available'))
   let g:tsuquyomi_is_available = 0
-  echom '[Tsuquyomi] Shougo/vimproc.vim is not installed. Please install it.'
+  echom '[Tsuquyomi] vim8 is not installed. Please install it.'
```
Comment (prabirshrestha) [inline ftplugin/javascript.vim]: vim8 with jobs and lambda is not installed?

File: README.md [hunk 6]
```diff
@@ -27,8 +27,7 @@ Tsuquyomi does not provide syntax-highlight nor indentation. You can use the fol
 ## How to install
 Tsuquyomi requires the following:
 
-+ [Vim](http://www.vim.org/) (v7.4.0 or later)
-+ [Shougo/vimproc.vim](https://github.com/Shougo/vimproc.vim)
++ [Vim](http://www.vim.org/) (v8.0.0 or later)
```
Comment (prabirshrestha) [inline README.md]: v8 or later with jobs and lambda?

File: autoload/tsuquyomi/config.vim (hunk 2 above)
Comment (mattn) [inline autoload/tsuquyomi/config.vim]: vim8 should have lambda.

File: autoload/tsuquyomi/config.vim (hunk 3 above)
Comment (mattn) [inline autoload/tsuquyomi/config.vim]: vim8 should have job too.

File: ftplugin/typescript.vim (hunk 4 above)
Comment (mattn) [inline ftplugin/typescript.vim]: ditto

File: ftplugin/javascript.vim (hunk 5 above)
Comment (mattn) [inline ftplugin/javascript.vim]: ditto

File: README.md (hunk 6 above)
Comment (mattn) [inline README.md]: ditto

**Comment by mattn** [comment]
The error you mentioned should be fixed.

File: autoload/tsuquyomi/config.vim (hunk 1 above)
Comment (mattn) [inline autoload/tsuquyomi/config.vim]: Sorry, I'm not familier to neovim.

**Comment by mattn** [comment]
it seems that neovim have `jobstart`. https://neovim.io/doc/user/job_control.html

but neovim doesn't have equivalent to job_readraw.

**Comment by rhysd** [comment]
IMO: vimproc implementation should remain as optional for Vim7 environment because many distributions does not introduce Vim8 yet. Even in Ubuntu 16.04, 7.4.1689 is installed.

File: autoload/tsuquyomi/config.vim (hunk 1 above)
Comment (prabirshrestha) [inline autoload/tsuquyomi/config.vim]: we could do this in separate CR and wait for neovim to support lamdas.

Here is a an example https://github.com/prabirshrestha/async.vim

File: autoload/tsuquyomi/config.vim (hunk 2 above)
Comment (prabirshrestha) [inline autoload/tsuquyomi/config.vim]: my thought was vim could be compiled without lambda and jobs support. If not the case I'm ok

**Comment by prabirshrestha** [comment]
@mattn any updates on this? Based on the changes it does seem pretty easy to make it work using vimproc or jobs in vim8.

**Comment by cosminadrianpopescu** [comment]
I've tested it with VIM 8.0 and it works perfectly. No more vimproc dependency.

**Comment by ilazarte** [comment]
Timeline for master?  Using vim on Android via termux and it has 8 available, vimproc seems to be the weak link.

**Comment by mattn** [comment]
It seems to be difficult to separate vim8 or vimproc part briefly.

**Comment by mattn** [comment]
I rebased to master. Anyone, could you please try last change? I separeted code for vim8 and vimproc. If you use vim7 or neovim, it will still work using vimproc.

**Comment by mattn** [comment]
It seems working well.

tsuquyomi

File: autoload/tsuquyomi/config.vim [hunk 7]
```diff
@@ -18,14 +18,16 @@ let s:script_dir = expand('<sfile>:p:h')
...
 
   if !exists('g:tsuquyomi_is_available')
-    if !s:P.is_available()
-      " 1. vimproc installation check
+    if !s:is_vim8 && !s:P.is_available()
+      " 1. vimproc or vim8 installation check
       let g:tsuquyomi_is_available = 0
       call s:deleteCommand()
-      echom '[Tsuquyomi] Shougo/vimproc.vim is not installed. Please install it.'
+	  echom '[Tsuquyomi] Shougo/vimproc.vim or vim8 is not installed. Please install it.'
```
Comment (mattn) [inline autoload/tsuquyomi/config.vim]: For Me, replace tab to whitespaces.

File: autoload/tsuquyomi/config.vim (hunk 1 above)
Comment (mattn) [inline autoload/tsuquyomi/config.vim]: I separated code for vim8 and vimproc. So this function will be called only for vim8

**Comment by rhysd** [comment]
Thank you @mattn for the work! I'll try with my local environment (macOS)

**Comment by rhysd** [comment]
@mattn @Quramy I confirmed that both job and vimproc work fine.

**Comment by Quramy** [comment]
@mattn Thanks for your rebase, @rhysd thanks for your check the operations.

I've tried it with my local vim 8, but the following error is remaining...

```
Error detected while processing function tsuquyomi#hint[2]..tsuquyomi#tsClient#tsQuickinfo[2]..tsuquyomi#tsClient#sendCommandSyncResponse[3]..tsuquyomi#tsClient#sendRequest:
line    6:
E631: ch_sendraw(): write failed
```

Do you have some workarounds?

And the vimproc mode works well with my nvim.

**Comment by rhysd** [comment]
What is your environment? If macOS, I think I could try to reproduce it.

**Comment by Quramy** [comment]
- OS: mac 10.10.3

- vim: `vim --version` (installed with macports)

```
VIM - Vi IMproved 8.0 (2016 Sep 12, compiled Jan 27 2017 06:10:32)
MacOS X (unix) version
Included patches: 1-237
Compiled by MacPorts
Huge version without GUI.  Features included (+) or not (-):
+acl             +file_in_path    +mouse_sgr       +tag_old_static
+arabic          +find_in_path    -mouse_sysmouse  -tag_any_white
+autocmd         +float           +mouse_urxvt     -tcl
-balloon_eval    +folding         +mouse_xterm     +termguicolors
-browse          -footer          +multi_byte      +terminfo
++builtin_terms  +fork()          +multi_lang      +termresponse
+byte_offset     +gettext         -mzscheme        +textobjects
+channel         -hangul_input    +netbeans_intg   +timers
+cindent         +iconv           +num64           +title
-clientserver    +insert_expand   +packages        -toolbar
+clipboard       +job             +path_extra      +user_commands
+cmdline_compl   +jumplist        -perl            +vertsplit
+cmdline_hist    +keymap          +persistent_undo +virtualedit
+cmdline_info    +lambda          +postscript      +visual
+comments        +langmap         +printer         +visualextra
+conceal         +libcall         +profile         +viminfo
+cryptv          +linebreak       -python          +vreplace
-cscope          +lispindent      -python3         +wildignore
+cursorbind      +listcmds        +quickfix        +wildmenu
+cursorshape     +localmap        +reltime         +windows
+dialog_con      -lua             +rightleft       +writebackup
+diff            +menu            -ruby            -X11
+digraphs        +mksession       +scrollbind      -xfontset
-dnd             +modify_fname    +signs           -xim
-ebcdic          +mouse           +smartindent     -xpm
+emacs_tags      -mouseshape      +startuptime     -xsmp
+eval            +mouse_dec       +statusline      -xterm_clipboard
+ex_extra        -mouse_gpm       -sun_workshop    -xterm_save
+extra_search    -mouse_jsbterm   +syntax
+farsi           +mouse_netterm   +tag_binary
   system vimrc file: "/opt/local/etc/vimrc"
     user vimrc file: "$HOME/.vimrc"
 2nd user vimrc file: "~/.vim/vimrc"
      user exrc file: "$HOME/.exrc"
       defaults file: "$VIMRUNTIME/defaults.vim"
  fall-back for $VIM: "/opt/local/share/vim"
Compilation: /usr/bin/clang -c -I. -Iproto -DHAVE_CONFIG_H   -I/opt/local/include -DMACOS_X_UNIX  -pipe -Os -arch x86_64 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1
Linking: /usr/bin/clang   -L/opt/local/lib -Wl,-headerpad_max_install_names -arch x86_64 -o vim        -lm  -lncurses -liconv -lintl -framework Cocoa
```

**Comment by Quramy** [comment]
In my local env, it seems `job_start` in `s:startTssVim8` creates a process, but  `:TsuStatusServer` says `'dead'`.

the above `job_start` tries to run `node "/Users/yosuke/.vim/bundle/tsuquyomi/node_modules/typescript/bin/tsserver"`.

**Comment by Quramy** [comment]
Ok, I got it.
In my env, using locally installed tsserver. 
The `job_start` arguments is `node  "/Users/yosuke/.vim/bundle/tsuquyomi/node_modules/typescript/bin/tsserver"`, which includes double quotation and it seems that this is an invalid argument for `job_start` func.

And I found that stripping quotation from the argument creates a valid tsserver process.

**Comment by Quramy** [comment]
@mattn Thanks for your great PR!

**Comment by mattn** [comment]
Thanks!

File: autoload/tsuquyomi/tsClient.vim [hunk 8]
```diff
@@ -67,17 +70,57 @@ function! tsuquyomi#tsClient#startTss()
...
+  endif
+  let l:cmd = substitute(tsuquyomi#config#tsscmd(), '\\', '\\\\', 'g')
+  try
+    let s:tsq['job'] = job_start(l:cmd)
+    let s:tsq['channel'] = job_getchannel(s:tsq['job'])
+
+    let out =  ch_readraw(s:tsq['channel'])
+    let st = tsuquyomi#tsClient#statusTss()
+    if !g:tsuquyomi_tsserver_debug
+      if err != ''
```
Comment (gurdiga) [inline autoload/tsuquyomi/tsClient.vim]: @mattn I’m just wondering: where does `err` come from?

File: autoload/tsuquyomi/tsClient.vim [hunk 9]
```diff
@@ -67,17 +70,57 @@ function! tsuquyomi#tsClient#startTss()
...
+function! s:startTssVim8()
+  if type(s:tsq['job']) == 8 && job_info(s:tsq['job']).status == 'run'
+    return 'existing'
+  endif
+  let l:cmd = substitute(tsuquyomi#config#tsscmd(), '\\', '\\\\', 'g')
+  try
+    let s:tsq['job'] = job_start(l:cmd)
+    let s:tsq['channel'] = job_getchannel(s:tsq['job'])
+
+    let out =  ch_readraw(s:tsq['channel'])
```
Comment (gurdiga) [inline autoload/tsuquyomi/tsClient.vim]: I’ve updated `tsuquyomi` yesterday and found that vim now opens TS files with a couple seconds delay. After stepping through it with `echom` statements, I found that his `ch_readraw` call takes more than 2 seconds. 😶

I’m completely new to Vimscript, but it still looks like `out` variable is not used, nor `st` on the next line. So, just to try out, it I’ve deleted lines 81–89 and also the `try`/`catch` wrapping, and after that `tsuquyomi` still seems to work, and the 2 seconds startup delay seems to be gone. 🎉 🕺 

I’m sure there is more to the picture that I could see in my first long hour starring at this code, so, what do you guys think may be a reasonable approach to get rid of the startup delay? 🤔

**Pull request merged**